packages/react-link/src/components/Link/Link.tsx

```diff
@@ -3,16 +3,18 @@ import { useLink } from './useLink';
 import { useLinkStyles } from './useLinkStyles';
 import { renderLink } from './renderLink';
 import type { LinkProps } from './Link.types';
+import type { ForwardRefComponent } from '@fluentui/react-utilities';
 
 /**
  * A Link is a reference to data that a user can follow by clicking or tapping it.
  */
-export const Link = React.forwardRef<HTMLAnchorElement | HTMLButtonElement, LinkProps>((props, ref) => {
+export const Link: ForwardRefComponent<LinkProps> = React.forwardRef((props, ref) => {
   const state = useLink(props, ref);
 
   useLinkStyles(state);
 
   return renderLink(state);
-});
+  // Work around some small mismatches in inferred types which don't matter in practice
+}) as ForwardRefComponent<LinkProps>;
```

Comment:
In the past I've had some bad experiences with the type inference for ISelection with DetailsList where new Selection has to be case to ISelection and somehow the constructor required params didn't match.  When I see a type declared and an cast at the end, it causes a bit of worry.  I think more detail about which inferred types are mismatching would help here.

Reply:
(Side note: see #13411 for issues with Selection constructor types...multiple people tried to fix it and couldn't come up with anything that fully worked 😢)

Reply:
Will this cast be required for all components with polymorphic `as` ?

Reply:
Currently yes, because I haven't been able to figure out a way to make it work with the way `forwardRef` manipulates the returned props type. You or others are welcome to try too if you want.

Reply:
It's possible that we could also improve things if we can find a way to simplify our prop definition helper types, but that's its own challenge (which I'm wondering if we'll need to do at some point since typescript 4.3+ [doesn't like our types](https://github.com/microsoft/fluentui/issues/19507), and even besides that they're very expensive to evaluate)

Reply:
First on `ComponentProps` we're removing `ref` from typings from Props and then after here at `ForwardRefComponent` we're pretty much going back to adding `ref` just to compensate the fact we've removed it in the first place.

![image](https://user-images.githubusercontent.com/5483269/135245059-8e5b68dd-7c04-41eb-aa01-d8aec182f78b.png)

Instead of removing `ref` from props just to only get it back in with `ForwardRefComponent` we simply stick with it and create a helper type (`RefFromProps`) to extract it, this way we can define components by simply using `forwardRef`

![image](https://user-images.githubusercontent.com/5483269/135248573-233b3a69-9e76-4871-9c42-9faa440642f8.png)

Reply:
I think the theoretical issue with having `ref` in component props is that technically it's not always present unless you use `forwardRef` on the exported component (which most components do, but a handful don't).

Reply:
That's a valid point, I wasn't thinking about another use case besides `forwardRef` as it's the way we're implementing internally, but it's valid that this might not be true for users

Reply:
Seems like this casting is a common problem when dealing with union types on Props [across libraries that try to follow this similar pattern](https://github.com/radix-ui/primitives/blob/dbefd647297bd6594577c3fc253dc874fbe11438/packages/react/polymorphic/src/polymorphic.test.tsx#L16-L20)

I guess there's no good solution besides accepting for now that a casting is required.

Reply:
@ecraig12345 Turns out that this is a problem already solved in the internals of `@types/react`, they updated the implementation of `PropsWithoutRef` which was breaking union types when `forwardRef` was invoked, all we need to do is upgrade our `@types/react` (and probably `@types/react-dom`) and this will be solved

Reply:
@bsunderhus Is the fix you're looking at present in `@types/react` 16, or only 17?

In [latest 16.x](https://unpkg.com/browse/@types/react@16.14.16/index.d.ts) I'm seeing this:
```ts
    type PropsWithoutRef<P> =
        // Just Pick would be sufficient for this, but I'm trying to avoid unnecessary mapping over union types
        // https://github.com/Microsoft/TypeScript/issues/28339
        'ref' extends keyof P
            ? Pick<P, Exclude<keyof P, 'ref'>>
            : P;
```

vs. this in [latest 17.x](https://unpkg.com/browse/@types/react@17.0.27/index.d.ts):
```ts    type PropsWithoutRef<P> =
        // Pick would not be sufficient for this. We'd like to avoid unnecessary mapping and need a distributive conditional to support unions.
        // see: https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types
        // https://github.com/Microsoft/TypeScript/issues/28339
        P extends any ? ('ref' extends keyof P ? Pick<P, Exclude<keyof P, 'ref'>> : P) : P;
```

Reply:
Meant to add: if it's only in 17, we unfortunately can't take that because we haven't upgraded to 17 yet. Though we could potentially try submitting a similar fix for 16 to DefinitelyTyped.